### PR TITLE
chore: Use Azure SDK for auth

### PR DESCRIPTION
**Pull request by helayoty**
- Remove Azure German Cloud.
- Move the auth functionality into a separate folder and use Azure SDK instead.

old code path(s):
- https://github.com/virtual-kubelet/azure-aci/blob/master/provider/aci.go#L190-L260
- https://github.com/virtual-kubelet/azure-aci/blob/master/client/auth.go
- https://github.com/virtual-kubelet/azure-aci/blob/master/provider/acsCredential_test.go

### CODE REVIEW

File: pkg/auth/auth_test.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import (
+	"io/ioutil"
+	"os"
+	"testing"
+)
+
+const cred = `
+{
+    "cloud":"AzurePublicCloud",
+    "tenantId": "72f988bf-86f1-41af-91ab-2d7cd011db47",
```
Comment (Fei-Guo) [inline pkg/auth/auth_test.go:12]: Is this a real tenantId? If both tenantId and subscriptionId are faked. please add a comment in the same line saying they are faked strings.

File: pkg/auth/auth.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+		if subscriptionID := os.Getenv("AZURE_SUBSCRIPTION_ID"); subscriptionID != "" {
+			c.AuthConfig.SubscriptionID = subscriptionID
+		}
+	}
+
+	resource := c.Cloud.Services[cloud.ResourceManager].Endpoint
+	options := &arm.ClientOptions{}
+
+	if cs, ok := options.Cloud.Services[cloud.ResourceManager]; ok {
```
Comment (Fei-Guo) [inline pkg/auth/auth.go:142]: In line 140, I think options are empty, so will line 142 be meaningless? You can defer the cleanup if this change only contains code move.

**Review by Fei-Guo — COMMENTED** [review]
Please post the original code link in the change description, so we know this PR solely contains code moves. If not, please highlight the changed parts.

File: pkg/auth/auth.go (hunk 2 above)
Comment (helayoty) [reply pkg/auth/auth.go:142]: For some reason, if we didn't add line 140, the `options.Cloud.Services[cloud.ResourceManager]` will return an empty string!

File: pkg/auth/auth_test.go (hunk 1 above)
Comment (helayoty) [reply pkg/auth/auth_test.go:12]: The old code used to use a real tenant and subscription IDs! Planning to get rid of them in the next phase.

Comment (helayoty) [reply pkg/auth/auth_test.go]: Replaced.

File: pkg/auth/auth.go (hunk 2 above)
Comment (helayoty) [reply pkg/auth/auth.go]: Removed and replaced by implicit import for arm package.

File: pkg/auth/auth.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+		c.AuthConfig = auth
+	}
+
+	if aksCredFilepath := os.Getenv("AKS_CREDENTIAL_LOCATION"); aksCredFilepath != "" {
+		c.AKSCredential, err = NewAKSCredential(aksCredFilepath)
+		if err != nil {
+			return err
+		}
+
+		if c.AKSCredential != nil {
```
Comment (Fei-Guo) [inline pkg/auth/auth.go:86]: You may not need the nil check since if err == nil, the aksCredential should not be nil.

File: pkg/auth/auth.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+	return nil
+}
+
+// Authentication represents the Authentication file for Azure.
+type Authentication struct {
+	ClientID             string `json:"clientId,omitempty"`
+	ClientSecret         string `json:"clientSecret,omitempty"`
+	SubscriptionID       string `json:"subscriptionId,omitempty"`
+	TenantID             string `json:"tenantId,omitempty"`
+	UseUserIdentity      bool   `json:"useUserIdentity,omitempty"`
```
Comment (Fei-Guo) [inline pkg/auth/auth.go:148]: UseUserIdentity flag can be derived from the length of the clientID field. You may not need this field in the struct.

File: pkg/auth/auth.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+	UserIdentityClientId string `json:"userIdentityClientId,omitempty"`
+}
+
+// NewAuthenticationFromFile returns an Authentication struct from file path
+func (a *Authentication) NewAuthenticationFromFile(filepath string) error {
```
Comment (Fei-Guo) [inline pkg/auth/auth.go:153]: If you don't plan to expose this method to be called in other modules, make it an internal method (newAuthenticationFromFile).

File: pkg/auth/auth.go [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+	ClientSecret           string `json:"aadClientSecret"`
+	ResourceGroup          string `json:"resourceGroup"`
+	Region                 string `json:"location"`
+	VNetName               string `json:"vnetName"`
+	VNetResourceGroup      string `json:"vnetResourceGroup"`
+	UserAssignedIdentityID string `json:"userAssignedIdentityID"`
+}
+
+// NewAKSCredential returns an aksCredential struct from file path
+func NewAKSCredential(p string) (*aksCredential, error) {
```
Comment (Fei-Guo) [inline pkg/auth/auth.go:199]: Can this be an internal method?  rename p-> filepath?

File: pkg/auth/auth.go (hunk 3 above)
Comment (helayoty) [reply pkg/auth/auth.go:86]: The old design was taken into consideration if customers want to run the provider on unmanaged AKS.

File: pkg/auth/auth.go (hunk 4 above)
Comment (helayoty) [reply pkg/auth/auth.go:148]: This is how the code determines the value [here](https://github.com/virtual-kubelet/azure-aci/pull/299/files/7563cac2f7a24ceafb6f8a6859dc0e610e01bb1a#diff-0ccbf9a2a7d709b87dbf2f30cc730528c6299d537bf721300d49502cfe521c0aR113).

Comment (Fei-Guo) [reply pkg/auth/auth.go:148]: I mean you don't need a field. You can expose a method
func (a Authentication) UseUserIdentity bool {
    return len(c.AuthConfig.ClientID) == 0
}
whenever you want to check this boolean.

File: pkg/auth/auth.go (hunk 3 above)
Comment (Fei-Guo) [reply pkg/auth/auth.go:86]: I am checking the method "NewAKSCredential", &cred cannot be nil if err is nil.

File: pkg/auth/auth.go (hunk 6 above)
Comment (helayoty) [reply pkg/auth/auth.go]: done

File: pkg/auth/auth.go (hunk 5 above)
Comment (helayoty) [reply pkg/auth/auth.go]: done.

File: pkg/auth/auth.go (hunk 4 above)
Comment (helayoty) [reply pkg/auth/auth.go]: done.

File: pkg/auth/auth.go (hunk 3 above)
Comment (helayoty) [reply pkg/auth/auth.go]: done.

File: pkg/auth/auth.go [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+			c.AuthConfig.ClientSecret = clientSecret
+		}
+
+		if userIdentityClientId := os.Getenv("VIRTUALNODE_USER_IDENTITY_CLIENTID"); userIdentityClientId != "" {
+			c.AuthConfig.UserIdentityClientId = userIdentityClientId
+		}
+
+		isUserIdentity := len(c.AuthConfig.ClientID) == 0
+
+		if isUserIdentity {
```
Comment (Fei-Guo) [inline pkg/auth/auth.go:116]: You can further remove "isUserIdentity" variable and add a comment instead. Feel free to do this in next change.

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**